lib/syntaxtree.gi

```diff
@@ -0,0 +1,831 @@
+                  , rhs := compiler.compile(expr.rhs));
+    end;
+    compiler.T_UNB_GVAR := expr -> expr;
+
+    compiler.T_ASS_LIST := compile_record;
+    compiler.T_ASSS_LIST := compile_record;
+
+    compiler.T_ASS_LIST_LEV := compile_record;
+    compiler.T_ASSS_LIST_LEV := compile_record;
+
+    compiler.T_UNB_LIST := compile_record;
+
+    compiler.T_ASS_REC_NAME := function(expr)
+        return rec( type := expr.type
+                  , rnam := expr.rnam
+                  , rhs := compiler.compile(expr.rhs) );
+    end;
+    compiler.T_ASS_REC_EXPR := compile_record;
+
+    compiler.T_UNB_REC_NAME := expr -> expr;
+    compiler.T_UNB_REC_EXPR := compile_record;
+
+    compiler.T_ASS_POSOBJ := compile_record;
+    compiler.T_ASSS_POSOBJ := compile_record;
+    compiler.T_ASS_POSOBJ_LEV := compile_record;
+    compiler.T_ASSS_POSOBJ_LEV := compile_record;
+    compiler.T_UNB_POSOBJ := compile_record;
+
+    compiler.T_ASS_COMOBJ_NAME := function(expr)
+        return rec( type := expr.type
+                  , rnam := expr.rnam
+                  , rhs := compiler.compile(expr.rhs) );
+    end;
+    compiler.T_ASS_COMOBJ_EXPR := compile_record;
+    compiler.T_UNB_COMOBJ_NAME := expr -> expr;
+    compiler.T_UNB_COMOBJ_EXPR := compile_record;
+
+    # Compiler.evaluates arguments lazily
+    compiler.T_INFO := unsupported;
+
+    compiler.T_ASSERT_2ARGS := unsupported;
+    compiler.T_ASSERT_3ARGS := unsupported;
+
+    compiler.T_EMPTY := expr -> expr;
+
+    # Options
+    compiler.T_PROCCALL_OPTS := unsupported;
+
+    # HPC-GAP's atomic statement (what about readwrite/readonly?)
+    compiler.T_ATOMIC := unsupported;
+
+    # A function expression
+    compiler.T_FUNC_EXPR := function(expr)
+        return rec( type := "T_FUNC_EXPR"
+                  , argnams := expr.argnams
+                  , narg := expr.narg
+                  , locnams := expr.locnams
+                  , nloc := expr.nloc
+                  , stats := compiler.compile(expr.stats) );
+    end;
+
+    compiler.T_OR := compile_record;
+    compiler.T_AND := compile_record;
+    compiler.T_NOT := compile_record;
+    compiler.T_EQ := compile_record;
+    compiler.T_NE := compile_record;
+    compiler.T_LT := compile_record;
+    compiler.T_GE := compile_record;
+    compiler.T_GT := compile_record;
+    compiler.T_LE := compile_record;
+    compiler.T_IN := compile_record;
+    compiler.T_SUM := compile_record;
+    compiler.T_AINV := compile_record;
+    compiler.T_DIFF := compile_record;
+    compiler.T_PROD := compile_record;
+    compiler.T_INV := compile_record;
+    compiler.T_QUO := compile_record;
+    compiler.T_MOD := compile_record;
+    compiler.T_POW := compile_record;
+
+    # These come from literals
+    # TODO: Maybe make the component names uniform
+    compiler.T_TRUE_EXPR := expr -> expr;
+    compiler.T_FALSE_EXPR := expr -> expr;
+    compiler.T_INTEXPR := expr -> expr;
+    compiler.T_INT_EXPR := expr -> expr;
```

Comment:
Off-topic, but: we probably should rename `T_INTEXPR` and `T_INT_EXPR`: one is for small/immediate integers, the other for large/big integers. Right now, they are super easy to confuse, and no way to tell which is which.

More off topic: we refer to "large integers" in some places, but it seems the rest of the world calls them "big integers" or "bigints", so perhaps we should also switch to that terminology.